### PR TITLE
More validation about cgroup-parent for systemd

### PR DESCRIPTION
**Pull request by hqhq**
Follow up https://github.com/docker/docker/pull/19154

Signed-off-by: Qiang Huang h.huangqiang@huawei.com

### CODE REVIEW

File: daemon/daemon_unix.go [hunk 1]
```diff
@@ -426,8 +448,8 @@ func verifyPlatformContainerSettings(daemon *Daemon, hostConfig *containertypes.
 		}
 	}
 	if hostConfig.CgroupParent != "" && daemon.usingSystemd() {
-		// CgroupParent for systemd cgroup should be named as "xxx.slice"
-		if len(hostConfig.CgroupParent) <= 6 || !strings.HasSuffix(hostConfig.CgroupParent, ".slice") {
+		// CgroupParent for systemd cgroup should be a valid slice named as "xxx.slice" or "xxx-yyy.slice"
+		if !isValidSlice(hostConfig.CgroupParent) {
 			return warnings, fmt.Errorf("cgroup-parent for systemd cgroup should be a valid slice named as \"xxx.slice\"")
```
Comment (cyphar) [inline daemon/daemon_unix.go]: Maybe update this error message? Just to say "invalid slice name".

Comment (hqhq) [inline daemon/daemon_unix.go]: Updated, thanks.

File: daemon/daemon_unix.go [hunk 2]
```diff
@@ -426,9 +448,9 @@ func verifyPlatformContainerSettings(daemon *Daemon, hostConfig *containertypes.
 		}
 	}
 	if hostConfig.CgroupParent != "" && daemon.usingSystemd() {
-		// CgroupParent for systemd cgroup should be named as "xxx.slice"
-		if len(hostConfig.CgroupParent) <= 6 || !strings.HasSuffix(hostConfig.CgroupParent, ".slice") {
-			return warnings, fmt.Errorf("cgroup-parent for systemd cgroup should be a valid slice named as \"xxx.slice\"")
+		// CgroupParent for systemd cgroup should be a valid slice named as "xxx.slice" or "xxx-yyy.slice"
+		if !isValidSlice(hostConfig.CgroupParent) {
+			return warnings, fmt.Errorf("cgroup-parent for systemd cgroup should be a valid slice name")
```
Comment (thaJeztah) [inline daemon/daemon_unix.go]: should this output the actual name that's used? e.g. "invalid slice name: %q"

Comment (calavera) [inline daemon/daemon_unix.go]: agreed

**Comment by hqhq** [comment]
@thaJeztah @calavera Updated, thanks.

**Comment by jfrazelle** [comment]
can we get a unit test for that function :innocent:

**Comment by cyphar** [comment]
I just figured out that `@` is also an invalid slice character in systemd:

```patch
From 41115d46c896af8a77b5a5a51c4767f447f62284 Mon Sep 17 00:00:00 2001
From: Aleksa Sarai <asarai@suse.com>
Date: Thu, 4 Feb 2016 09:10:39 +1100
Subject: [PATCH 2/2] daemon: systemd: filter @ in slice names

Slices don't take arguments, so the '@' symbol is not a valid slice
name.

Signed-off-by: Aleksa Sarai <asarai@suse.com>
---
 daemon/daemon_unix.go | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/daemon/daemon_unix.go b/daemon/daemon_unix.go
index c3f12d20449d..f882edd09b4c 100644
--- a/daemon/daemon_unix.go
+++ b/daemon/daemon_unix.go
@@ -367,8 +367,8 @@ func isValidSlice(sliceName string) bool {
        if len(sliceName) <= len(suffix) || !strings.HasSuffix(sliceName, suffix) {
                return false
        }
-       // Should not contain any path-separators.
-       if strings.Contains(sliceName, "/") {
+       // Should not contain any path-separators or @ arguments.
+       if strings.Contains(sliceName, "/") || strings.Contains(sliceName, "@") {
                return false
        }
 
-- 
2.7.0
```

**Comment by hqhq** [comment]
@cyphar Thanks, I added check for '@', but I think there must be other symbols that are not valid characters in systemd, and should not check all of them, using valid path/file names for `--cgroup-parent` should be a premiss, other than that, we can leave it to systemd when trying to create the slices.

@jfrazelle Added a unit test, thanks.

**Comment by calavera** [comment]
LGTM

**Comment by cyphar** [comment]
@hqhq Fair enough. This looks good.

**Comment by thaJeztah** [comment]
ping @anusha-ragunathan could you have a quick look?

**Comment by anusha-ragunathan** [comment]
For slice validation rules, refer https://github.com/systemd/systemd/blob/fdb4ee00f022863ceee923b196f9c6dd536db9e2/src/basic/unit-name.c#L55 
Digits, letters and a few special characters are allowed (depending on whether the unit name is plain, instance or template).  
 
Is there a point of partial validation of certain characters (eg. exclude @ for unit names). IMO, we either implement it fully, or let systemd do it.

**Comment by cyphar** [comment]
@anusha-ragunathan I agree we should implement validation fully. Thanks for providing the relevant code snippet. :P

@hqhq Can you take a look at the code @anusha-ragunathan posted and re-implement it here? I don't think we need to worry about doing this for libcontainer, since the `expandSlices` code is only really meant to deal with creating consistent cgroup paths (validation is a side issue).

**Comment by hqhq** [comment]
@anusha-ragunathan Thanks for your link, now I think it's better to leave validations to systemd, current validation for basic "xxx.slice" mode would be enough on Docker side. So I'm gonna close this PR and open a new one to add docs part only.

**Comment by hqhq** [comment]
https://github.com/docker/docker/pull/20310 picked docs part of this PR.

**Pull request closed without merging**